### PR TITLE
HPCC-12691 Add ESP method to report Component Status

### PR DESCRIPTION
**Pull request by wangkx**
The ws_machine.GetComponentStatus is added to report
Component Status which may be reported by Nagios or
other source. Nagios may report the status through
ws_machine.UpdateComponentStatus. If Component Status
is reported from multiple sources, the information
from the multiple sources is merged into one report.

Signed-off-by: wangkx kevin.wang@lexisnexis.com

### CODE REVIEW

**Comment by wangkx** [comment]
@garonsky please review.

File: esp/services/ws_machine/componentstatus.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    virtual int getComponentStatusID() = 0;
+    virtual const char* getTimeReportedStr() = 0;
+    virtual __int64 getTimeReported() = 0;
+    virtual const char* getEndPoint() = 0;
+    virtual const char* getComponentType() = 0;
+    virtual const int getComponentTypeID() = 0;
+    virtual IEspStatusReport* getStatusReport() = 0;
+    virtual IArrayOf<IEspComponentStatus>& getComponentStatusList() = 0;
+    virtual void setComponentStatus(IArrayOf<IConstComponentStatus>& StatusList) = 0;
+    virtual void mergeCachedComponentStatus(IESPComponentStatusInfo& statusInfo) = 0;
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.hpp]: Should some of this functions be protected/private?  Isn't mergeCachedComponentStatus should only be called internally by the class?

File: esp/services/ws_machine/componentstatus.cpp [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+#include "esp.hpp"
+#include "ws_machine_esp.ipp"
+#include "ws_machine.hpp"
+
+static unsigned csCacheMinutes = 30; //30 minutes as default
+static MapStringTo<int> componentTypeMap;
+static MapStringTo<int> componentStatusTypeMap;
+static unsigned componentTypeIDCount;
+
+class CESPComponentStatusInfo : public CInterface, implements IESPComponentStatusInfo
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.cpp]: Shouldn't this have a header file?

File: esp/services/ws_machine/componentstatus.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    #ifdef SMCLIB_EXPORTS
+        #define COMPONENTSTATUS_API __declspec(dllexport)
+    #else
+        #define COMPONENTSTATUS_API __declspec(dllimport)
+    #endif
+#else
+    #define COMPONENTSTATUS_API
+#endif
+
+class IESPComponentStatusInfo : public IInterface
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.hpp]: Should this be an .ipp file?

File: esp/services/ws_machine/componentstatus.cpp [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        if (csCacheMinutes > 0)
+        { //A status report should be expired if it was reported >csCacheMinutes ago.
+            CDateTime timeNow;
+            timeNow.setNow();
+            timeNow.adjustTime(-csCacheMinutes);
+            expireTimeStamp = timeNow.getSimple();
+        }
+    };
+
+    virtual const char* getReporter() { return reporter.get(); };
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.cpp]: These should probably be inline.

File: esp/services/ws_machine/componentstatus.cpp [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        {
+            IPropertyTree& statusType = statusTypes->query();
+            const char* name = statusType.queryProp("@name");
+            if (name && *name)
+                componentStatusTypeMap.setValue(name, statusType.getPropInt("@id"));
+        }
+
+        if (componentStatusTypeMap.count() < 1)
+        {
+            componentStatusTypeMap.setValue("normal", 1);
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.cpp]: Why 2 formats for string?

File: esp/services/ws_machine/componentstatus.cpp [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+############################################################################## */
+
+#include "jiface.hpp"
+#include "componentstatus.hpp"
+#include "jlib.hpp"
+#include "esp.hpp"
+#include "ws_machine_esp.ipp"
+#include "ws_machine.hpp"
+
+static unsigned csCacheMinutes = 30; //30 minutes as default
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.cpp]: Should these be class members?

File: esp/services/ws_machine/componentstatus.cpp [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+    int componentStatusID; //the worst component status in the system
+    int componentTypeID; //the worst component status in the system
+    StringAttr componentStatus; //the worst component status in the system
+    StringAttr componentType; //the worst component status in the system
+    StringAttr endPoint; //the worst component status in the system
+    StringAttr timeReportedStr; //the worst component status in the system
+    __int64 timeReported; //the worst component status in the system
+    Owned<IEspStatusReport> componentStatusReport; //the worst component status in the system
+
+    const char* formatTimeStamp(time_t tNow, StringAttr& out)
```
Comment (garonsky) [inline esp/services/ws_machine/componentstatus.cpp]: Should probably be a static method as there is no state.

File: esp/services/ws_machine/componentstatus.cpp (hunk 4 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.cpp]: Ok, I will change them to inline.

File: esp/services/ws_machine/componentstatus.cpp (hunk 6 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.cpp]: No. The csCacheMinutes is set by static void init(). It should only be initialized once.

File: esp/services/ws_machine/componentstatus.hpp (hunk 1 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.hpp]: No. The Get functions are designed for external access to the data members of the class. Both setComponentStatus() and mergeCachedComponentStatus() are called by class CComponentStatusFactory which caches the CESPComponentStatusInfo.

File: esp/services/ws_machine/componentstatus.cpp (hunk 2 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.cpp]: I can move to header file.

File: esp/services/ws_machine/componentstatus.cpp (hunk 7 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.cpp]: ok

File: esp/services/ws_machine/componentstatus.hpp (hunk 3 above)
Comment (wangkx) [inline esp/services/ws_machine/componentstatus.hpp]: I can.

**Comment by wangkx** [comment]
@garonsky I made the changes based on your review. Please review again.

**Comment by garonsky** [comment]
Looks good.

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to merge.

**Comment by richardkchapman** [comment]
@jakesmith @afishbeck Any efficiency concerns?

**Comment by garonsky** [comment]
A proper configuration would have the 'Monitoring ESP' outside a 'working' cluster.  This is the setup I am working with Ops to set up in our internal environment.  In a properly configured data center environment, the ESP handling monitoring should be dedicated for the task.  Depending on configuration settings, the Monitoring ESP can also monitor multiple clusters.

Additionally, the number and frequency of events is configurable by the Nagios admin using our Nagios tools and general Nagios configuration settings.  The load on the esp should never be at any substantial level to impact performance.  Also, the ESPs that receive the notification events are configurable as well, so not every ESP in a cluster would have to be involved in monitoring.

Use HPCC and Nagios together requires some prerequisite knowledge and configuration experience with Nagios in order properly configure a monitoring solution.

**Comment by richardkchapman** [comment]
I'm not just worried about the load on the ESP - I'm also concerned about any load that the ESP might put on dali, or in the clusters it is monitoring

**Comment by wangkx** [comment]
Those two ESP methods do not communicate with dali. @garonsky may address the concern about the load in the clusters it is monitoring.

**Comment by garonsky** [comment]
There's no question monitoring adds some load.  

Periodic checks could normally be expected to occur in a data center monitoring environment somewhere on the order of every 5 to 15 minutes.  This is configured by the Admin, customized for his or her needs.  

Nagios allows checks as frequently as 1x per minute.  Usually, after a fixed number of failures the checks stop. This, like most everything else, is configurable.

The number of checks that could occur is variable.  However it's not unreasonable that each node may have a dozen or more types of checks that are performed.

# of Esp Component Update Calls  = (types of  checks performed) X (check frequency) X (number of Nodes being monitored)

For example on the 190 cluster with roughly 100 nodes, with a 5 minute checks interval, and monitoring about 10 metrics on each node, we have the following:

10 x 12 x 100 = 12,000 events per hour or about 200 events a minute in the worse case. In practice this would really be a cluster stop/start that was unscheduled, without alerts being squelched.

Of course, this is only addressing the load on the ESP, the load on the nodes being monitored is important to consider, but it is outside the scope of this pull request.

**Pull request merged**